[lldb] Add support for unique target ids

qxy11 opened the pull request:
### Summary
Add support for unique target ids per Target instance. This is needed for upcoming changes to allow debugger instances to be shared across separate DAP instances for child process debugging. We want the IDE to be able to attach to existing targets in an already runny lldb-dap session, and having a unique ID per target would make that easier.

Each Target instance will have its own unique id, and uses a function-local counter in `TargetList::CreateTargetInternal` to assign incremental unique ids. 

### Tests
Added several unit tests to test basic functionality, uniqueness of targets, and target deletion doesn't affect the uniqueness.
```
bin/lldb-dotest -p TestDebuggerAPI
```

qxy11 committed:
Add support for unique target ids

qxy11 committed:
Merge branch 'main' into unique-target-id

JDevlieghere commented:
In the context of MCP support, we [have been discussing](https://github.com/llvm/llvm-project/issues/158676) adding a unique debugger identifier. If we're going to come up with something unique, it may be nice to make it globally unique, even across debuggers.

jimingham commented on lldb/source/Target/TargetList.cpp:
```diff
@@ -344,6 +346,8 @@ Status TargetList::CreateTargetInternal(Debugger &debugger,
   if (!target_sp)
     return error;
 
+  target_sp->m_target_unique_id = ++g_target_unique_id;
```
Yes, this seems much simpler to me, and I see no reason to delay assigning the Target's unique ID.

jimingham commented:
> In the context of MCP support, we [have been discussing](https://github.com/llvm/llvm-project/issues/158676) adding a unique debugger identifier. If we're going to come up with something unique, it may be nice to make it globally unique, even across debuggers.

I'm confused.  Debugger's already have a unique ID:

Debugger's already have a unique, monotonically increasing id:

```
Debugger::Debugger(lldb::LogOutputCallback log_callback, void *baton)
    : UserID(g_unique_id++),

```

That's what gets returned by Debugger::GetID().

The problem here is that we refer to Targets just by their index in the target list.  

But I wonder if it wouldn't be better to change THAT, and have the targets listed by this unique ID instead.  It's kind of bogus that if you have three targets, and you remove the middle one, that changes how you refer to the targets in all the command interfaces for them.

jimingham commented:
I guess you mean by "Giving the debugger a unique ID" you mean "unique across all instances of lldb past or present"...

I'm not sure why you really want that, for all the ones currently extant the combo of {lldb pid, debugger ID} would do.  Why do you care about not overlapping with defunct Debuggers?

In any case, whatever you call the debugger, {DebuggerID, TargetID} would still be universally unique if the debugger was and the Target ID was unique within a Debugger.  So it seems unnecessary work to make the Target's universally unique.

ashgti commented:
Having something like a session identifier could also work for the mcp use case, e.g. `lldb://session/<uuid>/debugger/0/target/5` where we generate the uuid when the MCP server is started.

That provides us with enough information to distinguish two lldb instances.

walter-erquinigo commented:
> That's what gets returned by Debugger::GetID().

@jimingham , what about having `GetID()` return the unique ID global to all debugger instances. And then have a new method `GetTargetIndex()` that would return the same value that `GetID` is returning now and modify existing references to point to it.

qxy11 committed:
Set unique ID in Target constructor

qxy11 committed:
Use atomic counter for target id

Summary:
Also use lldb::user_id_t throughout instead of uint32_t. It seems to be the custom around the codebase to use this type for unique ids rather than a uint32_t

qxy11 committed:
use lldb::user_id_t instead of uint32_t

jimingham commented:
> @jimingham , thanks for the info. It seems you edited my comment instead of replying :)
> 
> I any case, I'm fine with the existing direction of this patch w.r.t. simply adding a global ID. Any particular feedback about the current implementation?

The code seems fine.  You never say what this ID is Unique with respect to.  If it were me I'd call the API GetGloballyUniqueID since this is guaranteeing a unique-ness beyond the Target's natural container - the Debugger.  Or you could just more explicitly document the behavior you are implementing.  You are also asserting that this is a globally Unique ID, but you don't actually test that, since all your tests have only one debugger.  You should probably also make a couple Debuggers and assert that their "UniqueID's" also don't overlap.

walter-erquinigo commented:
@qxy11: Jim gave you good feedback. Try to apply that and I'll do a review

qxy11 committed:
Add another unit test and change GetUniqueID -> GetGloballyUniqueID

qxy11 committed:
Fix formatting

qxy11 committed:
Remove unnecessary formatting

walter-erquinigo commented on lldb/include/lldb/lldb-defines.h:
```diff
@@ -96,6 +96,7 @@
 #define LLDB_INVALID_QUEUE_ID 0
 #define LLDB_INVALID_CPU_ID UINT32_MAX
 #define LLDB_INVALID_WATCHPOINT_RESOURCE_ID UINT32_MAX
+#define LLDB_INVALID_TARGET_ID 0
```
```suggestion
#define LLDB_INVALID_GLOBALLY_UNIQUE_TARGET_ID 0
```

walter-erquinigo commented on lldb/test/API/python_api/debugger/TestDebuggerAPI.py:
```diff
@@ -294,3 +294,145 @@ def test_version(self):
         self.assertEqual(class_str, property_str)
+
+    def test_find_target_with_unique_id(self):
+        """Test SBDebugger.FindTargetWithUniqueID() functionality."""
```
don't forget to rename this as well

walter-erquinigo commented on lldb/test/API/python_api/debugger/TestDebuggerAPI.py:
```diff
@@ -294,3 +294,145 @@ def test_version(self):
+
+        # Find the target using its unique ID
+        unique_id = target.GetGloballyUniqueID()
+        self.assertNotEqual(unique_id, 0)
```
don't use 0 here. Instead, expose the constant for the invalid thread id to the SB API and use it here

walter-erquinigo commented on lldb/test/API/python_api/debugger/TestDebuggerAPI.py:
```diff
@@ -294,3 +294,145 @@ def test_version(self):
+                    found.GetGloballyUniqueID(), target.GetGloballyUniqueID()
+                )
+
+        # Clean up
+        lldb.SBDebugger.Destroy(debugger1)
+        lldb.SBDebugger.Destroy(debugger2)
```
add these as a tearDown callback after you create the debuggers, otherwise they won't be destroyed automatically after a failed assert IIRC

walter-erquinigo reviewed (changes requested):
pretty nice. Almost there

qxy11 committed:
FindTargetWithUniqueID -> FindTargetByGloballyUniqueID and test case cleanup

walter-erquinigo commented on lldb/include/lldb/API/SBDebugger.h:
```diff
@@ -359,6 +359,9 @@ class LLDB_API SBDebugger {
   lldb::SBTarget FindTargetWithFileAndArch(const char *filename,
                                            const char *arch);
 
+  /// Find a target with the specified unique ID
```
```suggestion
  /// Find a target with the specified unique ID.
```

walter-erquinigo reviewed (approved):
fancy

qxy11 committed:
Add missing periods.
